x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -231,16 +235,19 @@ public LoggingAuditTrail(Settings settings, ClusterService clusterService, Threa
     }
 
     @Override
-    public void authenticationSuccess(String requestId, String realm, User user, RestRequest request) {
+    public void authenticationSuccess(String requestId, Authentication authentication, RestRequest request) {
         if (events.contains(AUTHENTICATION_SUCCESS) && eventFilterPolicyRegistry.ignorePredicate()
-                .test(new AuditEventMetaInfo(Optional.of(user), Optional.of(realm), Optional.empty(), Optional.empty())) == false) {
+                .test(new AuditEventMetaInfo(
+                        Optional.of(authentication.getUser()),
+                        Optional.of(effectiveRealmName(authentication)),
+                        Optional.empty(),
+                        Optional.empty())) == false) {
             final StringMapMessage logEntry = new LogEntryBuilder()
                     .with(EVENT_TYPE_FIELD_NAME, REST_ORIGIN_FIELD_VALUE)
                     .with(EVENT_ACTION_FIELD_NAME, "authentication_success")
-                    .with(REALM_FIELD_NAME, realm)
```

Comment:
`authentication_success` is different because it returned the realm inside the `realm` field instead of `user.realm` field. It almost makes sense, but the problem is that the `user.name` could reside in the lookup realm, which is not audited. Also the authn realm name could be `_es_api_key` which is even less helpful.

I feel this change is borderline. Happy to discuss it.

Reply:
We have quite a few realm related fields:
* `realm`
* `user.realm`
* `user.run_by.realm`
* `user.run_as.realm`

Is there a valid situation that all four of them would have different meanings? Or should we get rid of one of them, e.g. `realm`, completely?

Reply:
> Is there a valid situation that all four of them would have different meanings?

The simple answer is no. In general, they each appear in different event types.

`user.run_as.realm` only appears for `run_as_granted` and `run_as_denied` events. `user.run_by.realm` and `realm` fields are not present for such events. `user.realm` shows the authenticating realm name.

For `access_granted` and `access_denied` events, `user.*` fields always represent the attributes of the effective user. If the user is "run as", the `user.run_by.*` fields show the attributes of the authenticated user.

`realm` is only present for `authentication_success` and `authentication_failed` events, but my suggestion, in this PR, is to get rid of it for `authentication_success`. It is still useful for `authentication_failed` events, when no other `user.*` fields are present, because there is no user yet (the user is built after authentication succeeded).

---

To clarify, `authentication_success` currently (before this PR) includes only the `realm` field from your list, but it also contains the `user.name` attribute of the user. `realm` contains the realm that did the authentication, and `user.name` contains the effective authentication (it could be the looked up username, `authentication_success` is generated after a possible `run_as_granted`). In this PR I propose we remove the `realm` field, and rely on the strategy we use for `access_granted` authz events, where the `user.*` fields always represent the attributes of the effective user and `user.run_by.*` the attributes of the authenticated user.
Although I believe it is a good change, I am pushing for it now because, in order to better support the API Keys, the `realm` would contain the effective realm name, because the authenticating realm name is not recorded on API key creation. So we cannot easily maintain the current significance of the `realm` field, and I reckon it's actually better if we remove the field altogether and rely on `user.*` and `user.run_by.*` paradigm.

Reply:
Sorry does `... because the authenticating realm name is not recorded on API key creation` mean that we do not currently log API key creation, i.e. we are logging anything that happens after auth flow?

For `authentication_success`, I agree it feels more natural to use the same convention as `access_grant` etc instead of `authentication_fail`. My only concern is BWC, i.e. how much friction would it cause to end users by changing the field name in a minor release.

Reply:
This concerns me, assuming I understand it correctly.
1. The BWC issue of removing a field in a minor. I don't think we should do that
2. Under this change, for `authentication_success` with _run-as_ the realm that did the _authentication_ would be in `user.run_by.realm`, but for all other cases it would be in `user.realm`.

I would propose (but I'm open to other options)
- Keep the change (below) for `withAuthentication(auth)` so that we now audit the `user.realm` and `user.run_by.*`
- But also keep the `realm` field to indicate the realm against which the authentication occurred (so it's always in a single, known field).
- For API Key authentication, `realm` should be either `_es_api_key` or `null` because authentication didn't occur against a realm

Reply:
Thanks Tim, your understanding matches what I meant and the proposed option sounds good to me.

As a parting comment, the `realm` field now contains redundant information and I would propose we remove it in the next major.

Reply:
Pushed 8aa90686381135ce53e56bfd827ae6467b67bfb4

Reply:
> As a parting comment, the realm field now contains redundant information and I would propose we remove it in the next major.

I'm not sure - but we can discuss. I have thoughts around the degree of consistency we want between `realm_authentication_failed` and `authentication_success`  vs `access_granted` and `authentication_success`.

I presume that `realm_authentication_failed` will always have a `realm` field, because there's no other place to record the realm, so even if we remove it from `authentication_success` it will still exist in the file (/index). Is it better to retain it on `authentication_success` as well, for clarity? I don't know.